lib/tokenizers.js

```diff
@@ -225,52 +243,61 @@ async function _tokenizerFromRecord({record}) {
-  // 3. Store the tokenizer record with pending state.
-  const collection = database.collections['tokenizer-tokenizer'];
-  const now = Date.now();
-  const meta = {created: now, updated: now};
-  const tokenizer = {
-    id: capabilityAgent.id,
-    secret: base64url.encode(secret),
-    state: 'pending'
-  };
-  let record = {
-    meta,
-    tokenizer
-  };
   try {
-    const result = await collection.insertOne(record, database.writeOptions);
-    record = result.ops[0];
-  } catch(e) {
-    if(!database.isDuplicateError(e)) {
-      throw e;
+    // 1. Generate a random secret.
+    const secret = await randomBytesAsync(32);
+    // 2. Generate capability agent from handle and secret.
+    const handle = 'primary';
+    const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+    // 3. Store the tokenizer record with pending state.
+    const collection = database.collections['tokenizer-tokenizer'];
+    const now = Date.now();
+    const meta = {created: now, updated: now};
+    const tokenizer = {
+      id: capabilityAgent.id,
+      secret: base64url.encode(secret),
+      state: 'pending'
+    };
+    let record = {
+      meta,
+      tokenizer
+    };
+    try {
+      const result = await collection.insertOne(record, database.writeOptions);
+      record = result.ops[0];
+    } catch(e) {
+      if(!database.isDuplicateError(e)) {
+        throw e;
+      }
+      throw new BedrockError(
+        'Duplicate tokenizer.',
+        'DuplicateError', {
+          public: true,
+          httpStatusCode: 409
+        }, e);
     }
-    throw new BedrockError(
-      'Duplicate tokenizer.',
-      'DuplicateError', {
-        public: true,
-        httpStatusCode: 409
-      }, e);
-  }
 
-  // 4. Create a keystore and HMAC key for the tokenizer.
-  const keystore = await kms.createKeystore({
-    controller: capabilityAgent.id, referenceId: 'primary'
-  });
-  const keystoreAgent = kms.getKeystoreAgent(
-    {capabilityAgent, keystoreId: keystore.id});
-  const key = await keystoreAgent.generateKey({type: 'hmac'});
-
-  // 5. Ready tokenizer for use by adding keystore and HMAC key to its record.
-  await _readyTokenizer({tokenizer, keystore, key});
-
-  // mark any ready tokenizer as current
-  await _markTokenizerAsCurrent();
+    // 4. Create a keystore and HMAC key for the tokenizer.
+    const keystore = await kms.createKeystore({
+      controller: capabilityAgent.id, referenceId: 'primary'
+    });
+
+    const keystoreAgent = kms.getKeystoreAgent(
+      {capabilityAgent, keystoreId: keystore.id});
+    const key = await keystoreAgent.generateKey({type: 'hmac'});
+
+    // 5. Ready tokenizer for use by adding keystore and HMAC key to its record.
+    await _readyTokenizer({tokenizer, keystore, key});
+
+    // mark any ready tokenizer as current
+    await _markTokenizerAsCurrent();
+  } catch(cause) {
+    const error = new BedrockError(
+      'Error creating keystore for tokenizer.', 'UnknownError',
```

Comment:
Why is `UnknownError` used here?

Reply:
Let's use `OperationError`. An operation error occurs when there's some operation-specific problem that occurred -- and that specific problem will be expressed by the `cause`.

Reply:
Changed, thanks.